File: acts-sys/src/definitions/algebra.rs
pub use ffi::*;

#[cxx::bridge(namespace = "Acts")]
mod ffi {
    unsafe extern "C++" {
        include!("Acts/Definitions/Algebra.hpp");
        include!("acts-sys/include/definitions/algebra.hpp");

        type Vector2;
        type Vector3;
        type Vector4;
        type SquareMatrix2;
        type SquareMatrix3;
        type SquareMatrix4;
        type Translation2;
        type Translation3;
        type RotationMatrix2;
        type RotationMatrix3;
        type AngleAxis3;
        type Transform2;
        type Transform3;

        #[namespace = "acts_sys::ffi"]
        #[rust_name = "transform_equivalent_tolerance"]
        fn s_transformEquivalentTolerance() -> f64;

        #[namespace = "acts_sys::ffi"]
        fn make_transform3(
            rotation_x: f64,
            rotation_y: f64,
            rotation_z: f64,
            translation_x: f64,
            translation_y: f64,
            translation_z: f64,
        ) -> UniquePtr<Transform3>;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_transform_equivalent_tolerance() {
        let tol = transform_equivalent_tolerance();
        assert_eq!(tol, 1e-9);
    }
}

File: acts-sys/src/definitions/algebra.cpp
#include "acts-sys/include/definitions/algebra.hpp"
#include <Eigen/Geometry>

namespace acts_sys {
namespace ffi {

std::unique_ptr<Acts::Transform3>
make_transform3(double rotation_x, double rotation_y, double rotation_z,
                double translation_x, double translation_y,
                double translation_z) {
  Eigen::AngleAxisd rot_x(rotation_x, Eigen::Vector3d::UnitX());
  Eigen::AngleAxisd rot_y(rotation_y, Eigen::Vector3d::UnitY());
  Eigen::AngleAxisd rot_z(rotation_z, Eigen::Vector3d::UnitZ());

  Eigen::Matrix3d rotation;
  rotation = rot_z * rot_y * rot_x;

  Acts::Transform3 tform = Acts::Transform3::Identity();
  tform.linear() = rotation;
  tform.translation() =
      Eigen::Vector3d(translation_x, translation_y, translation_z);

  return std::make_unique<Acts::Transform3>(tform);
}

} // namespace ffi
} // namespace acts_sys
